### PR TITLE
Add two functions for handling push durations

### PR DESCRIPTION
**Pull request by carolksun**
This PR adds a function to find the best duration unit for the duration of the 
pushes in a push def as well as a function to find the first non-empty state 
start time. The best duration unit is the maximum time unit that the majority 
of the durations can use. The function to find the first non-empty state start 
time finds the start and end time for that push such that that the start time is 
the start time of the first non-empty state and the end time is the start time of
the last state.

### CODE REVIEW

File: web/src/app/components/duration.utils.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
+import {step189_2020} from '../../proto/step189_2020';
+
+export interface DurationItem {
+  start: number|Long;  // nsec time of the first non-empty state
+  end: number|Long;    // nsec time of the last state
```
Comment (razvanm) [inline web/src/app/components/duration.utils.ts:5]: `startNsec` and `endNsec` would be more explicit about the unit,

File: web/src/app/components/duration.utils.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  seconds: NANO_TO_SECONDS,
+  minutes: NANO_TO_MINUTES,
+  hours: NANO_TO_HOURS,
+  days: NANO_TO_DAYS
+};
+
+/**
+ * Finds the unit of time that best describes the majority of the durations of
+ * the pushInfos.
+ * @param pushInfos Array of pushes for a single push def
```
Comment (razvanm) [inline web/src/app/components/duration.utils.ts:23]:
```suggestion
 * the pushInfos.
 *
 * @param pushInfos Array of pushes for a single push def
```

File: web/src/app/components/duration.utils.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+/**
+ * Finds the start time of the first non-empty state and the ned time of the
+ * pushInfo.
+ * @param pushInfo Array of pushes for a single push def
+ * @return a DurationItem that has a start of the time of the first non-empty
+ *     stage and an end of the time of the last state
+ */
+export function findStartandEnd(pushInfo: step189_2020.IPushInfo): DurationItem|
```
Comment (razvanm) [inline web/src/app/components/duration.utils.ts:79]: How about calling this `findDuration`?

File: web/src/app/components/duration.utils.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    if (value > max) {
+      max = value;
+      bestUnit = key;
+    }
+  }
+  return bestUnit;
+}
+
+/**
+ * Finds the start time of the first non-empty state and the ned time of the
```
Comment (razvanm) [inline web/src/app/components/duration.utils.ts:73]:
```suggestion
 * Finds the start time of the first non-empty state and the end time of the
```

File: web/src/app/components/duration.utils.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+ * pushInfo.
+ * @param pushInfo Array of pushes for a single push def
```
Comment (razvanm) [inline web/src/app/components/duration.utils.ts:75]: Is this accurate? The argument seems to be a single push.

File: web/src/app/components/duration.utils.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+ * @return a DurationItem that has a start of the time of the first non-empty
+ *     stage and an end of the time of the last state
```
Comment (razvanm) [inline web/src/app/components/duration.utils.ts:77]: We should also indicate when we returned unknown.

File: web/src/app/components/duration.utils.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+ * Finds the unit of time that best describes the majority of the durations of
+ * the pushInfos.
+ *
+ * @param pushInfos Array of pushes for a single push def
+ * @return one of [seconds, minutes, hours, days]
+ */
+export function findDurationUnit(pushInfos: step189_2020.IPushInfo[]): string {
+  const unitCounter:
+      {[unit: string]:
+           number} = {seconds: 0, minutes: 0, hours: 0, days: 0, years: 0};
```
Comment (razvanm) [inline web/src/app/components/duration.utils.ts:30]: A bit nicer form:

```
  const unitCounter: {[unit: string]: number} = {
    seconds: 0,
    minutes: 0,
    hours: 0,
    days: 0,
    years: 0,
  };
```

File: web/src/app/components/duration.utils.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  });
+  let bestUnit = '';
+  let max = 0;
+  for (const [key, value] of Object.entries(unitCounter)) {
+    if (value > max) {
+      max = value;
+      bestUnit = key;
+    }
+  }
+  return bestUnit;
```
Comment (razvanm) [inline web/src/app/components/duration.utils.ts:70]: As discussed on Google Chat, some line breaks would help. Same in the other function.

File: web/src/app/components/duration.utils.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  if (firstStateStart === -1) {
+    return;
+  }
+
+  const endTime = states[states.length - 1].startTimeNsec;
+  if (!endTime) {
+    return;
+  }
+
+  return {startNsec: firstStateStart, endNsec: endTime} as DurationItem;
```
Comment (yinggao) [inline web/src/app/components/duration.utils.ts:112]:
```suggestion
  return { startNsec: firstStateStart, endNsec: endTime } as DurationItem;
```

**Pull request merged**